components/mediators/datamapper/org.wso2.carbon.mediator.datamapper.engine/src/main/java/org/wso2/carbon/mediator/datamapper/engine/input/readers/XMLInputReader.java

```diff
@@ -0,0 +1,512 @@
+    private static final String PROPERTIES_KEY = "properties";
+    private static final String ATTRIBUTES_KEY = "attributes";
+    private static final String TYPE_KEY = "type";
+
+    private static final Log log = LogFactory.getLog(XMLInputReader.class);
+
+    /* Reference of the InputXMLMessageBuilder object to send the built JSON message */
+    private InputBuilder messageBuilder;
+
+    /* JSON schema of the input message */
+    private Schema inputSchema;
+
+    /* Name and NamespaceURI of the currently processing XML element */
+    private String localName;
+    private String nameSpaceURI;
+
+    /* JSON Builder to build the respective JSON message */
+    private JSONBuilder jsonBuilder;
+
+    /* Iterator for the Attribute elements */
+    private Iterator<OMAttribute> it_attr;
+
+    /* JSON schema for input message */
+    private Map jsonSchema;
+
+    /**
+     * Constructor
+     *
+     * @throws IOException
+     */
+    public XMLInputReader() throws IOException {
+        this.jsonBuilder = new JSONBuilder();
+    }
+
+    /**
+     * Read, parse the XML and notify with the output JSON message
+     *
+     * @param input          XML message InputStream
+     * @param inputSchema    Schema of the input message
+     * @param messageBuilder Reference of the InputXMLMessageBuilder
+     * @throws ReaderException Exceptions in the parsing stage
+     */
+    @Override
+    public void read(InputStream input, Schema inputSchema, InputBuilder messageBuilder) throws ReaderException {
+
+        this.messageBuilder = messageBuilder;
+        this.inputSchema = inputSchema;
+
+        OMXMLParserWrapper parserWrapper = OMXMLBuilderFactory.createOMBuilder(input);
+        OMElement root = parserWrapper.getDocumentElement();
+        this.jsonSchema = getInputSchema().getSchemaMap();
+
+        try {
+            XMLTraverse(root, null, jsonSchema);
+            jsonBuilder.writeEndObject();
+            writeTerminateElement();
+        } catch (IOException | JSException | SchemaException | InvalidPayloadException e) {
+            throw new ReaderException("Error while parsing XML input stream. " + e.getMessage());
+        }
+
+    }
+
+    /**
+     * This method will perform a Depth First Search on the XML message and build the json message
+     *
+     * @param omElement       initially the root element will be passed-in
+     * @param prevElementName name of the previous element only if the previous element was an array, a null otherwise
+     * @param jsonSchemaMap   reduced json input schema map that is applicable to this level
+     * @return the name of the previous element if the element was an array element, null otherwise
+     * @throws IOException
+     * @throws ReaderException
+     * @throws SchemaException
+     * @throws JSException
+     * @throws InvalidPayloadException
+     */
+    public String XMLTraverse(OMElement omElement, String prevElementName, Map jsonSchemaMap)
```

Comment:
by convention, java method name starts with lower case. so change it so `xmlTraverse(...)`